src/edu/csus/ecs/pc2/ui/DefineCustomInputValidatorPane.java

```diff
@@ -0,0 +1,350 @@
+            inputValidatorCommandTextField.setPreferredSize(new Dimension(200, 25));
+            inputValidatorCommandTextField.setMinimumSize(new Dimension(200, 25));
+            inputValidatorCommandTextField.setColumns(20);
+            inputValidatorCommandTextField.setText("");
+            inputValidatorCommandTextField.setToolTipText(null);
+            
+            inputValidatorCommandTextField.addKeyListener(new KeyAdapter() {
+                public void keyPressed(KeyEvent event) {
+                    enableUpdateButton();
+                    updateInputValidatorPaneComponents();
+                    //command has changed; mark any previous execution invalid
+                    try {
+                        ((InputValidatorPane) getParentPane()).setCustomInputValidatorHasBeenRun(false);
+                        ((InputValidatorPane) getParentPane()).setCustomInputValidationStatus(InputValidationStatus.NOT_TESTED);
+                    } catch (ClassCastException e) {
+                        getController().getLog().severe("Parent of DefineCustomInputValidatorPane is not an InputValidatorPane; not supported: " + e); 
+                        JOptionPane.showMessageDialog(null, "Internal error: parent of DefineCustomInputValidatorPane is not an InputValidatorPane; "
+                                + "\nSee logs and please report this to the PC2 Development team (pc2@ecs.csus.edu)",
+                                                "Internal Error", JOptionPane.ERROR_MESSAGE); 
+                    }
+                }
+            });
+        }
+        return inputValidatorCommandTextField;
+    }
+    
+
+    /**
+     * Calls enableUpdateButton() in the grandparent EditProblemPane if that pane exists.
+     * What that method does is check the various GUI components to determine if anything about the
+     * problem configuration has changed, and if so enables the Update button on the EditProblemPane, allowing
+     * the user to save the problem changes.
+     */
+    private void enableUpdateButton() {
+        JPanePlugin parent = getParentPane();
+        if (parent != null && parent instanceof InputValidatorPane) {
+            final JPanePlugin grandParent = ((InputValidatorPane)parent).getParentPane();
+            if (grandParent != null && grandParent instanceof EditProblemPane) {
+                SwingUtilities.invokeLater( new Runnable() {
+                    public void run () {
+                        ((EditProblemPane)grandParent).enableUpdateButton();
+                    }
+               });
+            } else {
+                System.err.println ("Internal error: grandparent of DefineCustomInputValidatorPane is not an EditProblemPane; cannot enable Add/Update button"
+                        + "\nPlease report this problem to the PC2 Development Team (pc2@ecs.csus.edu");
+                getController().getLog().severe("grandparent of DefineCustomInputValidatorPane is not an EditProblemPane");

```

Comment:
Should log message and displayed messages be identical?

Reply:
I don't see why this should necessarily be the case; displayed messages are generally for the user sitting at the console, while log messages are, generally speaking, more for the systems person who's going to pore through logs looking for detailed information which in fact might not make any sense to the average user at the console.

I will concede however that the systems person looking at the log might also like to see the "Internal error" part of the message, along with the "cannot enable Add/Update button"; I'll push that as a change to the PR.